### ISSUE TITLE
Create sign-in/up funnels in redash

### ISSUE THREAD

**Issue by philbooth** [mozilla/fxa-activity-metrics#10]
See the [feature doc](https://github.com/mozilla/fxa/tree/master/features/FxA-41-signin-funnel-metrics) for more information.

**Comment by philbooth** [comment]
In case anyone picks this one up, I should point out that the events documented in that feature doc begin in train 71. So right now, there is not much data to play with.

**Comment by davismtl** [comment]
Query started here : https://sql.telemetry.mozilla.org/queries/1485/source

**Comment by davismtl** [comment]
@philbooth just to confirm that the events in the table are either no longer good or will change names once train 71 lands and the events should be named like this: 
https://github.com/mozilla/fxa-auth-server/blob/7ef8d5d1c9e02f9bc119adf4b6356eb5716beb7d/docs/metrics-events.md

If so, then I think I may have completed most of the Login Funnel (see link in previous comment). It may just need a few adjustments once the data lands.

**Comment by davismtl** [comment]
Once the data has landed and the above query is validated, I propose to fork it to make the sign up funnel. Until then, I feel like this card is now blocked.

**Comment by davismtl** [comment]
@philbooth I updated the query with your latest event names. Can you review the query and the results it outputs and confirm that everything looks good for now?

**Comment by philbooth** [comment]
> Can you review the query and the results it outputs and confirm that everything looks good for now?

I think it is going to over-count certain steps in its current state and possibly under-count some others depending on what we want to include in this particular funnel.

For instance:
- Because of the `Have an account?` link on sign-up, it is possible for users to emit `flow.signin.engage` following `flow.signup.begin`. In those cases, we never see a `flow.signin.begin` event.
- Because of everyone's favourite feature, signin-from-signup, it is possible for `email.confirmation.sent` to be emitted following a `flow.signup.submit` rather than a `flow.signin.submit`. This was the reason for introducing the `flow.signin.attempt` and `flow.signup.attempt` events.
- Also the same problem exists for requests that originate from `force_auth`. Essentially, none of the server-side events are aware of the initiating view in the content server. They will count requests from all views.
- The `email.clicked` event is emitted for both sign-in confirmation and sign-up verification.
- Ditto `flow.complete`, it is emitted for all flows.

I think the general solution to fixing most of these is to `LEFT JOIN` each step of the funnel to its immediate follower, joining on `flow_id`. This will have the effect of treating each step as a filter against over-counting.

And because of signin-from-signup, we need to do one of two things to properly capture the `begin` events:
- Either wildcard the event name with `type LIKE 'flow.%.begin'`;
- Or explode the funnels so that we have separate ones for each different `begin` event, then for each different `engage` event following each different `begin` event, then for each different `attempt` event following each different `engage` event following each different `begin` event. (if that makes sense)

Or maybe we want to do both of those things and have the aggregate and specific funnels available to us.

Now I've written that out, I'm not sure it makes sense. I can also explain over Vidyo later if it's easier.

**Comment by davismtl** [comment]
Thanks for all the details. I'm going to reflect on this a bit. I'll see if I can outline all these funnels in a spreadsheet. It'll help visualize it and surface these types of scenarios.

**Comment by rfk** [comment]
Just trying to get my head around the status here, ahead of our sync-up tomorrow...

> I think it is going to over-count certain steps in its current state and possibly under-count
> some others depending on what we want to include in this particular funnel.

I definitely agree with this - there will be lots of flows that contain e.g. a "flow.completed" event but not a "flow.signin.engage" event, so we can't just count up the totals and show them side-by-side.

However...

> I think the general solution to fixing most of these is to LEFT JOIN each step of the 
> funnel to its immediate follower, joining on flow_id. 

This seems pretty scary, from a query complexity and performance perspective.  I'm hopeful we can find a way that doesn't involve lots of joints over the `flow_events` table.

I've been using the "signin unblock" flow graph as a testing ground, and I'm pretty happy with what I've got in the latest version of the query:

  https://sql.telemetry.mozilla.org/queries/1626/source

It's done in three parts, which I'll copy here for discussion.

1) Select the flows of interest.

First we narrow down our view to just the flows of interest:

```
WITH

  blocked_flows AS (
    -- For some reason, these flows are not int the flow_metadata table..?
    -- We have to take timestamp from the 'blocked' event itself.
    SELECT DISTINCT m.begin_time::DATE AS day, m.flow_id
    FROM flow_events e
    INNER JOIN flow_metadata m
    ON m.flow_id = e.flow_id
    WHERE e.type = 'customs.blocked'
    AND m.service = 'sync'
    ORDER BY 1, 2
  ),
```

This is important for the reasons Phil outlined above - we need to filter down to a specific type of flow, so that we know what events to expect.  Here we're looking only at flows for `service=sync` that contained a `customs.blocked` event.

2) Flag whether each flow contains the events of interest

Next, we check whether the relevant events are contained in each flow.

```
  funnel_event_flags AS (
    SELECT
      b.day,
      b.flow_id,
      COUNT(CASE WHEN e.type = 'flow.signin.submit' THEN true ELSE NULL END) > 0 AS did_signin_submit,
      COUNT(CASE WHEN e.type = 'account.login.sentUnblockCode' THEN true ELSE NULL END) > 0  AS did_unblock_send,
      COUNT(CASE WHEN e.type = 'account.login.confirmedUnblockCode' THEN true ELSE NULL END) > 0  AS did_unblock_confirm,
      COUNT(CASE WHEN e.type = 'account.login' THEN true ELSE NULL END) > 0  AS did_login,
      COUNT(CASE WHEN e.type = 'account.keyfetch' THEN true ELSE NULL END) > 0  AS did_keyfetch
    FROM blocked_flows AS b
    INNER JOIN flow_events AS e
    ON b.flow_id = e.flow_id
    GROUP BY 1, 2
    ORDER BY 1, 2
  ),
```

This groups by flow_id to tell us whether or not each individual event of interest occurred in each flow.  It's basically doing the same thing as Alex's query above, except I'm making a boolean rather than an integer.

Note that we might have rows here where e.g. `did_unblock_send` is False, but `did_login` is True.  Those would be people who choose to wait for the block to expire, rather than emailing their way around it.

3) Count by a conjunction of flags

This is where we enforce a strict funnel, asserting that the flow passed through each event in turn:

```
  funnel_event_counts_by_day AS (
    SELECT
      day,
      COUNT(flow_id) AS count_blocked,
      COUNT(CASE WHEN did_signin_submit THEN true ELSE NULL END) AS count_signin_submit,
      COUNT(CASE WHEN did_signin_submit AND did_unblock_send THEN true ELSE NULL END) AS count_unblock_send,
      COUNT(CASE WHEN did_signin_submit AND did_unblock_send AND did_unblock_confirm THEN true ELSE NULL END) AS count_unblock_confirm,
      COUNT(CASE WHEN did_signin_submit AND did_unblock_send AND did_unblock_confirm AND did_login THEN true ELSE NULL END) AS count_login,
      COUNT(CASE WHEN did_signin_submit AND did_unblock_send AND did_unblock_confirm AND did_login AND did_keyfetch THEN true ELSE NULL END) AS count_keyfetch
    FROM funnel_event_flags
    GROUP BY 1
    ORDER BY 1
  )
```

Note that we count each event only if all the previous events also occurred in that flow.

IIUC, this achieves the same effect that Phil was talking about, but without needing a lot of left joins.  It runs in about 40 seconds on the current dataset, which is pretty good!

Of course, we don't _have_ to filter like this.  As we saw with a previous version of the signin-unblock graph, it's also interesting to know how many people succeed after `customs.blocked` _without_ having to send an email.  But if we're going to talk about funnels in a strict percentage-dropoff-at-each-stage sense, I think an approach like the above will be necessary.

**Comment by rfk** [comment]
Here's an attempt at a sign-in funnel using the same technique:

 https://sql.telemetry.mozilla.org/queries/1674/source#2973

Not sure I completely trust it yet, e.g. there may be some flows that get an "account.login" without a preceeding "flow.signin.engage" due to missing client-side flow events.  But it seems like a step in the right direction.

**Comment by vladikoff** [comment]
from mtg: close. please create a new issue for new graphs to be added...
